store/store_k8s.go

```diff
@@ -0,0 +1,257 @@
+)
+
+type k8sStore struct {
+	cm configmap.ConfigMap
+}
+
+// NewK8sStore returns a Store implementation which uses
+// k8s configmaps to store data.
+func NewK8sStore(clusterID string) (Store, configmap.ConfigMap, error) {
+	k8sStore, cm, err := NewK8sStoreWithParams(
+		configmap.GetName(confgMapPrefix, clusterID),
+		configmap.DefaultK8sLockTimeout,
+		configmap.DefaultK8sLockAttempts * time.Second,
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+	return k8sStore, cm, nil
+}
+
+// NewK8sStoreWithParams returns a Store implementation which uses
+// k8s configmaps to store data. ConfigMap properties can be customized.
+func NewK8sStoreWithParams(
+	name string,
+	lockTryDuration time.Duration,
+	lockTimeout time.Duration,
+) (Store, configmap.ConfigMap, error) {
+	lockAttempts := uint((lockTryDuration / time.Second))
+	cm, err := configmap.New(
+		name,
+		nil,
+		lockTimeout,
+		lockAttempts,
+		0,
+		0,
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+	return &k8sStore{cm}, cm, nil
+}
+
+func (k8s *k8sStore) Lock(owner string) (*StoreLock, error) {
+	if err := k8s.cm.Lock(owner); err != nil {
+		return nil, err
+	}
+	return &StoreLock{owner: owner}, nil
+}
+
+func (k8s *k8sStore) LockWithKey(owner, key string) (*StoreLock, error) {
+	if err := k8s.cm.LockWithKey(owner, key); err != nil {
+		return nil, err
+	}
+	return &StoreLock{Key: key, owner: owner, lockedWithKey: true}, nil
+}
+
+func (k8s *k8sStore) Unlock(storeLock *StoreLock) error {
+	if storeLock.lockedWithKey {
+		return k8s.cm.UnlockWithKey(storeLock.Key)
+	}
+	return k8s.cm.Unlock()
+}
+
+func (k8s *k8sStore) IsKeyLocked(key string) (bool, string, error) {
+	return k8s.cm.IsKeyLocked(key)
+}
+
+func (k8s *k8sStore) CreateKey(key string, value []byte) error {
+	sanitizedKey := GetSanitizedK8sName(key)
+	err := k8s.cm.LockWithKey(string(value), sanitizedKey)
+	if err != nil {
+		logrus.Errorf("unable to lock with key %v", key)
+		return err
+	}
+	defer func() {
+		err := k8s.cm.UnlockWithKey(sanitizedKey)
+		if err != nil {
+			logrus.Infof("unable to unlock with key %v", key)
+		}
+	}()
+
+	data, err := k8s.cm.Get()
+	if err != nil {
+		return err
+	}
+
+	if _, ok := data[key]; ok {
+		return &StoreKeyExists{
+			Key:     key,
+			Message: "Use PutKey API",
+		}
+	}
+
+	if data == nil {
+		data = make(map[string]string)
+	}
+	data[key] = string(value)
+	return k8s.patchWithRetries(data)
+}
+
+func (k8s *k8sStore) PutKey(key string, value []byte) error {
+	data, err := k8s.cm.Get()
+	if err != nil {
+		return err
+	}
+
+	data[key] = string(value)
+	return k8s.patchWithRetries(data)
+}
+
+func (k8s *k8sStore) GetKey(key string) ([]byte, error) {
+	data, err := k8s.cm.Get()
+	if err != nil {
+		return nil, err
+	}
+
+	value, ok := data[key]
+	if !ok {
+		return nil, &StoreKeyDoesNotExist{
+			Key: key,
+		}
+	}
+
+	return []byte(value), nil
+}
+
+func (k8s *k8sStore) DeleteKey(key string) error {
+	sanitizedKey := GetSanitizedK8sName(key)
```

Comment:
I can't tell when we need to sanitize the key and when we don't. Also, not clear about when we need to lock. Please add comments to clarify. Also, functions could be named differently to distinguish if we are assuming that caller needs to have acquired a lock in some cases.

Reply:
CreateKey and DeleteKey were only use in vSphere and vSphere hostname had special characters and that was the reason it was added. These are existing functions. If you have some specific suggestions I can do them. The larger effort I'd prefer to defer to a later time since I am on a time crunch.